src/components/ReanimatedSwipeable.tsx

```diff
@@ -433,179 +461,196 @@ const Swipeable = forwardRef<SwipeableMethods, SwipeableProps>(
-          }
-        }
-
-        updateAnimatedEvent();
-      })
-      .onEnd(
-        (event: GestureStateChangeEvent<PanGestureHandlerEventPayload>) => {
-          handleRelease(event);
-        }
-      )
-      .onFinalize(() => {
-        dragStarted.value = false;
-      });
-
-    if (enableTrackpadTwoFingerGesture) {
-      panGesture.enableTrackpadTwoFingerGesture(enableTrackpadTwoFingerGesture);
-    }
-
-    panGesture.activeOffsetX([
-      -dragOffsetFromRightEdge,
-      dragOffsetFromLeftEdge,
-    ]);
-    tapGesture.shouldCancelWhenOutside(true);
+    const tapGesture = useMemo(
+      () =>
+        Gesture.Tap()
+          .shouldCancelWhenOutside(true)
+          .onStart(() => {
+            if (rowState.value !== 0) {
+              close();
+            }
+          }),
+      [close, rowState]
+    );
 
-    useImperativeHandle(ref, () => swipeableMethods.current, [
-      swipeableMethods,
-    ]);
+    const panGesture = useMemo(
+      () =>
+        Gesture.Pan()
+          .enabled(enabled !== false)
+          .enableTrackpadTwoFingerGesture(enableTrackpadTwoFingerGesture)
+          .activeOffsetX([-dragOffsetFromRightEdge, dragOffsetFromLeftEdge])
+          .onStart(() => {
+            updateRightElementWidth();
+          })
+          .onUpdate(
+            (event: GestureUpdateEvent<PanGestureHandlerEventPayload>) => {
+              userDrag.value = event.translationX;
+
+              const direction =
+                rowState.value === -1
+                  ? SwipeDirection.RIGHT
+                  : rowState.value === 1
+                  ? SwipeDirection.LEFT
+                  : event.translationX > 0
+                  ? SwipeDirection.RIGHT
+                  : SwipeDirection.LEFT;
+
+              if (!dragStarted.value) {
+                dragStarted.value = true;
+                if (rowState.value === 0 && onSwipeableOpenStartDrag) {
+                  runOnJS(onSwipeableOpenStartDrag)(direction);
+                } else if (onSwipeableCloseStartDrag) {
+                  runOnJS(onSwipeableCloseStartDrag)(direction);
+                }
+              }
+
+              updateAnimatedEvent();
+            }
+          )
+          .onEnd(
+            (event: GestureStateChangeEvent<PanGestureHandlerEventPayload>) => {
+              handleRelease(event);
+            }
+          )
+          .onFinalize(() => {
+            dragStarted.value = false;
+          }),
+      [
+        dragOffsetFromLeftEdge,
+        dragOffsetFromRightEdge,
+        dragStarted,
+        enableTrackpadTwoFingerGesture,
+        enabled,
+        handleRelease,
+        onSwipeableCloseStartDrag,
+        onSwipeableOpenStartDrag,
+        rowState,
+        updateAnimatedEvent,
+        updateRightElementWidth,
+        userDrag,
+      ]
+    );
 
-    panGesture.enabled(enabled !== false);
+    useImperativeHandle(ref, () => swipeableMethods.value, [swipeableMethods]);
```

Comment:
Why do we need to use `useImperativeHandle`?

Reply:
I'm not sure if there's another method for attaching methods to the ref of `ReanimatedSwipeable`.
Why do we want to avoid using it?

Reply:
I forgot that previous implementation was a class component. Either way, chances are that this might not work as expected since `swipeableMethods` is `SharedValue` and its reference does not change between renders. I think @j-piasecki will provide better explanation 😅

Reply:
The potential issue here is that shared values (and I think derived values as well) are preferentially stable between renders. In this case, the `useImperativeHandle` hook will execute exactly once as the swipeableMethods itself doesn't change - its value does.

Another problem is that the only time `useImperativeHandle` may be reevaluated is during render, but changes to shared values (and derived values, animated styles, etc.) don't cause a re-render. Due to this, changing the dependency array to `swipeableMethods.value` may or may not work, since it would need to depend on something else to trigger renders.

Reply:
I encountered a dependancy circle when working on memoizing the `swipeableMethods` object.
[`swipeableMethods` -> `animateRow` -> `dispatchEndEvents` -> `swipeableMethods`]
My solution for now is to remove `swipeableMethods` from the `dispatchEndEvents`, this unfortunately has to change the `onSwipeableOpen` and `onSwipeableClose` functions' signature from: 
```js
(
  direction: SwipeDirection.LEFT | SwipeDirection.RIGHT,
  swipeable: SwipeableMethods
) => void
``` 
to 
```js
(
  direction: SwipeDirection.LEFT | SwipeDirection.RIGHT,
) => void
```
I couldn't find an alternative solution for this problem, and I think there might not be a way around it.
Thankfully the now removed `swipeable: SwipeableMethods` argument is completely redundant, and can be replaced with the `swipeableRef` exposed via `useImperativeHandle`, 
this is also the way it currently works in the legacy `Swipeable` component.